### ISSUE TITLE
unsupported operand type(s) for +: 'NoneType' and 'str'

### ISSUE THREAD

**Issue by greysteil** [sdispater/poetry#1081]
- [x] I am on the [latest](https://github.com/sdispater/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/sdispater/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: MacOS 10.14.4
- **Poetry version**: 0.12.15
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: https://github.com/texastribune/tt-docker-base

## Issue

Running `pyenv exec poetry update django-object-actions --lock -vvv` on the linked project gives the following error on Poetry `0.12.15`. It works fine with `0.12.14`.

```
[TypeError]                                 
unsupported operand type(s) for +: 'NoneType' and 'str'  
                                                           
Exception trace:
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/cleo/application.py in run() at line 94
   status_code = self.do_run(input_, output_)
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/console/application.py in do_run() at line 88
   return super(Application, self).do_run(i, o)
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/cleo/application.py in do_run() at line 197
   status_code = command.run(input_, output_)
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/console/commands/command.py in run() at line 77
   return super(BaseCommand, self).run(i, o)
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/cleo/commands/base_command.py in run() at line 146
   status_code = self.execute(input_, output_)
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/cleo/commands/command.py in execute() at line 107
   return self.handle()
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/console/commands/update.py in handle() at line 41
   return installer.run()
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/installation/installer.py in run() at line 73
   self._do_install(local_repo)
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/installation/installer.py in _do_install() at line 165
   ops = solver.solve(use_latest=self._whitelist)
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/puzzle/solver.py in solve() at line 38
   packages, depths = self._solve(use_latest=use_latest)
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/puzzle/solver.py in _solve() at line 171
   self._package, self._provider, locked=locked, use_latest=use_latest
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/mixology/__init__.py in resolve_version() at line 7
   return solver.solve()
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/mixology/version_solver.py in solve() at line 79
   next = self._choose_package_version()
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/mixology/version_solver.py in _choose_package_version() at line 377
   version = self._provider.complete_package(version)
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/puzzle/provider.py in complete_package() at line 465
   package.name, package.version.text, extras=package.requires_extras
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/repositories/pool.py in package() at line 49
   package = repository.package(name, version, extras=extras)
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/repositories/legacy_repository.py in package() at line 270
   release_info = self.get_release_info(name, version)
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/repositories/pypi_repository.py in get_release_info() at line 272
   cached = self._get_release_info(name, version)
 /Users/greysteil/.pyenv/versions/3.7.3/lib/python3.7/site-packages/poetry/repositories/legacy_repository.py in _get_release_info() at line 363
   hashes.append(link.hash_name + ":" + hash)

update [--no-dev] [--dry-run] [--lock] [--] [<packages>]...
```

**Comment by chrahunt** [comment]
I was able to reproduce without `pyenv exec` using a pip-installed poetry 0.12.15 and can confirm it works with 0.12.14.

**Comment by chrahunt** [comment]
The problem is that the URLs for releases for [django-taggit](https://pypi.texastribune.org/simple/django-taggit/) do not include a hash. Previously if a hash was not set then `link.hash_name` would be `None` and a hash would simply not be included - which is true for django-taggit in the existing poetry.lock in the linked repo.

**Comment by neganp** [comment]
At first glance, it looks like [this change](https://github.com/sdispater/poetry/commit/3bc1c5aea99fecb20bc19cb057a910516d5a0e34#diff-82b23d5737d621bcdd9009dfa78a88f2R362) needs an `elif` clause to make sure that `link.hash_name` isn't `None`?

**Comment by chrahunt** [comment]
@neganp Yes that is where the change is needed if the goal is to preserve the original behavior. [PEP 503](https://www.python.org/dev/peps/pep-0503/) doesn't mandate a hash be in the URL, it would impact build reproducibility though, so if we accept it then we may want to trace a warning. @sdispater WDYT?

**Comment by mrosales** [comment]
This is a breaking regression for me as I'm on a private PyPi server that does not provide hashes. No packages can be installed and I had to revert to  `0.12.4` to restore functionality. Previous comments from @neganp and @chrahunt seem to correctly identify the location & cause of the issue. Are there any plans to fix this soon or is it something that help is wanted on?

**Comment by greysteil** [comment]
I'm not a Python dev, so I would *love* someone else to pick this up 🙏 :octocat:

**Comment by pappasam** [comment]
This bug has to do with using a private package index:

```toml
[[tool.poetry.source]]
name = "private-pypi"
url = "https://pypi.myurl.com/simple/"
```

```bash
# Fails with the error above
poetry add requests
```

Removing the private package index from the pyproject.toml resolves the issue

**Comment by mrosales** [comment]
Looks like this is fixed by 8b1afebb495bcaf7ffef5fcd59647ea6e783395c

**Comment by greysteil** [comment]
Looks that way, thanks!
